internal/daemon/api_logs.go

```diff
@@ -0,0 +1,312 @@
+	}
+}
+
+// logsResponse is a Response implementation to serve the logs in a custom
+// JSON Lines format.
+type logsResponse struct {
+	svcMgr serviceManager
+}
+
+func (r logsResponse) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	query := req.URL.Query()
+
+	services := query["services"]
+
+	followStr := query.Get("follow")
+	if followStr != "" && followStr != "true" && followStr != "false" {
+		response := statusBadRequest(`follow parameter must be "true" or "false"`)
+		response.ServeHTTP(w, req)
+		return
+	}
+	follow := followStr == "true"
+
+	numLogs := defaultNumLogs
+	nStr := query.Get("n")
+	if nStr != "" {
+		n, err := strconv.Atoi(nStr)
+		if err != nil {
+			response := statusBadRequest("n must be a valid integer")
+			response.ServeHTTP(w, req)
+			return
+		}
+		numLogs = n
+	}
+
+	// If "services" parameter not specified, fetch logs for all services.
+	if len(services) == 0 {
+		infos, err := r.svcMgr.Services(nil)
+		if err != nil {
+			response := statusInternalError("cannot fetch services: %v", err)
+			response.ServeHTTP(w, req)
+			return
+		}
+		services = make([]string, len(infos))
+		for i, info := range infos {
+			services[i] = info.Name
+		}
+	}
+
+	// Get log iterators by service (and close them when we're done).
+	itsByName, err := r.svcMgr.ServiceLogs(services, numLogs)
+	if err != nil {
+		response := statusInternalError("cannot fetch log iterators: %v", err)
+		response.ServeHTTP(w, req)
+		return
+	}
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Output format is JSON Lines, which doesn't have an official mime type,
+	// but "application/x-ndjson" is what most people seem to use:
+	// https://github.com/wardi/jsonlines/issues/9
+	w.Header().Set("Content-Type", "application/x-ndjson")
+	encoder := json.NewEncoder(w)
+	encoder.SetEscapeHTML(false)
+
+	// Use a buffered channel as a FIFO for keeping the latest numLogs logs if
+	// request "n" is set (the default).
+	var fifo chan servicelog.Entry
+	if numLogs > 0 {
+		fifo = make(chan servicelog.Entry, numLogs)
+	}
+	flushFifo := func() error { // helper to flush any logs in the FIFO
+		if numLogs <= 0 || len(fifo) == 0 {
+			return nil
+		}
+		var err error
+		for len(fifo) > 0 && err == nil {
+			err = encoder.Encode(jsonLog(<-fifo))
+		}
+		if err != nil {
+			logger.Noticef("error writing logs: %v", err)
+			return err
+		}
+		flushWriter(w)
+		return nil
+	}
+
+	// Close the streamLogs done channel when either the request is cancelled
+	// or we return from this handler.
+	returning := make(chan struct{})
+	defer close(returning)
+	done := make(chan struct{}, 1)
+	go func() {
+		select {
+		case <-req.Context().Done():
+		case <-returning:
+		}
+		done <- struct{}{}
+	}()
+
+	// Background goroutine to stream ordered logs: it sends parsed logs on
+	// logs channel, any error on errors channel.
+	logs := make(chan servicelog.Entry)
+	errors := make(chan error)
+	go streamLogs(itsByName, logs, errors, done)
```

Comment:
make streamLogs return an error then

```
go func() {
   errors <- streamLogs(itsByName, logs, ctx.Done())
}()
```

Reply:
Yeah, that's a bit nicer -- thanks. Done.